### ISSUE TITLE
moving focus after navigations

### ISSUE THREAD

**Issue by mayank99** [vercel/next.js#49386]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue

https://github.com/vercel/app-playground

### To Reproduce

Open any Nextjs app, click on any link, then inspect focus - it stays on the link or gets lost if link is no longer on the page. This is more obvious to see if you use keyboard to <kbd>Tab</kbd> to any link then press <kbd>Enter</kbd>.

### Describe the Bug

I've been keeping my eye on the App Router roadmap since it said "Accessibility". Now that it is stable, I see that it announces the page title on every navigation. This is a great first step! However, this is _not_ enough most of the times.

Consider a basic example of a documentation website containing a sidebar and a content area. Now if a keyboard user navigates to a different page using a link in the sidebar, the focus stays on that link even after navigation; any "skip-to-content" links that would normally be near the top of the page (and therefore quick to find) are now several tab stops away. And this is assuming the link is even on the page; if it's not then the focus will get _lost_, which is not ideal at all.

When you look at a regular (non-SPA) website, the browser will reset focus to the top of the page after navigation instead of leaving it somewhere in the middle. This makes sense because it's a new page. Since Nextjs is trying to recreate navigations using client-side JavaScript, I think it would similarly make sense to simulate browser behavior and reset focus (e.g. `document.body.focus()`). However, even this may not be ideal in complex dynamic apps with nested layouts, because a screen-reader/keyboard user would have to tediously traverse through the lots of elements to get to the part that actually changed.

The "best" solution will differ for different apps and even for different scenarios within the same app. In 2019, there was some very useful user testing done by Marcy Sutton about [accessible client-side routing techniques](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/#recommendations-finding-common-ground) - one of the main suggestions in there was to add a skip-link and move focus to it upon navigation. Today, nextjs does nothing to help with this.

### Expected Behavior

I do not expect Nextjs to magically make the router 100% accessible out-of-the-box, but I _do_ expect at least _some_ (or better, _all_) of the following:
- an option (per-link and global) to automatically reset focus to the top of page
- a hook/callback (per-navigation) to do manual focus management after each navigation (e.g. gatsby's `onRouteUpdate` / sveltekit's `afterNavigate`)
- a built-in skip-link component that, if present, nextjs can automatically focus after navigations
- provide info about the topmost nested element that actually changed
- provide info about whether the navigation was performed using mouse or keyboard
- provide the ability to disable next-route-announcer
- the ability to use regular browser navigations while keeping the "nice" things from `next/link` (such as prefetching)
- guidance and examples on how to use nextjs apis (including any new ones) to do accessible routing

Note: This issue is about _link_ navigations, but implementing some of these would also help with making _form_ navigations more accessible.

**Comment by mayank99** [comment]
i found one instance of `focus()`, but in my testing, the focus was still on the link, so i'm not sure about the heuristics under which it's called.
https://github.com/vercel/next.js/blob/173381a57239e25291116f12b522ccdc64e42624/packages/next/src/client/components/layout-router.tsx#L251

**Comment by diegohaz** [comment]
I can see the focus change on parallel routes. It seems to focus on the first focusable element of that nested layout or page when the route changes. I actually wish there was a way to control this behavior. For example, I'd like to disable this autofocus when I'm already handling the focus on the app (for example, when I'm opening or closing a dialog or changing tabs).

**Comment by mayank99** [comment]
@diegohaz You're right, there is some focus change _sometimes_, but I don't really understand the behavior and it also seems to be inconsistent.

For instance, navigating to/from https://app-dir.vercel.app/route-groups but navigating to/from https://app-dir.vercel.app/route-groups/clothing seems to keep focus on the link (I'm sure there is even different behavior in other cases).

https://github.com/vercel/next.js/assets/9084735/9b18ad60-2f5e-4040-b6f6-d990f19155e8

At minimum, the current behavior should be made consistent and the heuristics should be documented well.

In either case, I also agree with allowing disabling this behavior altogether. This could be a global config option or a customizable component in the root layout.
```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  focusAfterNavigation: 'manual'
};
module.exports = nextConfig;
```

Also would be useful to have more info for each navigation. This could be exposed through callbacks on the `<Link>` component. As an additional benefit, these callbacks would also serve as a reliable place for doing manual focus management (matters less for cases like dialogs which have their own triggers).
```jsx
<Link
  href='/somewhere'
  beforeNavigate={(info) => { console.log(document.activeElement); }}
  afterNavigate={(info) => { console.log(info.changedLayout) }}
>
  Go somewhere
</Link>
```

**Comment by owenniblock** [comment]
We're very interested in finding a good solution for this problem. Let me know if there's any way we can help here ❤️

**Comment by arti5m** [comment]
I'm following this too. In the meantime, I'm working around it by setting focus manually. 
I'd love to help out if you could point me in the right direction.

**Comment by mayank99** [comment]
We've crossed the 6 month anniversary of this issue.

@timneutkens I see you've edited the description to add a Linear link. I assume that means it's being tracked somehow. It would be nice to hear an official response and discuss a path forward, as this is a very serious issue which requires non-trivial workarounds.

**Comment by philwolstenholme** [comment]
I've just butted up against this issue when trying to do some route based tabs using a Next catch all segment and https://react-spectrum.adobe.com/react-aria/Tabs.html to try and have some tabs that respect URL state, e.g. `/page/tab-a` and `/page/tab-b` would both result in `page` being rendered but with either the first (a) or second (b) set of tabs open by default.

React Aria's focus management is great, but Next was resetting focus to the `body` element because I was using `next/link` via React Aria's [integration with client side routers](https://react-spectrum.adobe.com/react-aria/routing.html). When I abandoned my attempt to have the tabs be controlled by URL state then the issue was resolved as React Aria took over all the focus management.

**Comment by dzucconi** [comment]
To add to this: consider the common case of a modal dialog route (the Instagram profile grid is a good example of this) Closing out of dialog should yield focus back to the element that triggered the route, but due to Next's behavior it resets.

**Comment by Dupflo** [comment]
Just find a small trick, force focus on a tag element `<main/>` in my case to reset the tab index. Pretty efficient on footer link in my case

```ts
'use client'

import Link from 'next/link'
import type { MouseEvent } from 'react'

export function ResetFocusLink(props) {

  const handleClick = (e: MouseEvent<HTMLAnchorElement>) => {
    setTimeout(() => {
      const mainElement = document.querySelector('main')
      if (mainElement) {
        mainElement.setAttribute('tabindex', '-1')
        mainElement.focus()
        mainElement.removeAttribute('tabindex')
        window.scrollTo({ top: 0, behavior: 'instant' })
      }
    }, 100)
  }

  return <Link {...props} onClick={handleClick} />
}
